File: AgentforceSDK-ReactNative-Bridge/ios/Agentforce/Providers/BridgeViewProvider.swift
/*
 * Copyright (c) 2024-present, salesforce.com, inc. All rights reserved.
 *
 * Bridges the native AgentforceViewProviding protocol to React Native.
 * When enabled, delegates rendering of specified component types to a
 * registered React Native component via RCTRootView.
 */

import Foundation
import UIKit
import SwiftUI
import React
import AgentforceSDK

/// Implements AgentforceViewProviding by delegating to a React Native component.
/// Component types are registered synchronously from JS; rendering uses RCTRootView.
class BridgeViewProvider: AgentforceViewProviding {

    /// Set of component definition strings this provider handles (e.g. "copilot/richText")
    private var registeredTypes: Set<String> = []

    /// The React Native component name to render for matching types
    private var reactComponentName: String = ""

    /// Reference to the RCT bridge for creating root views
    private weak var bridge: RCTBridge?

    init(bridge: RCTBridge?) {
        self.bridge = bridge
    }

    /// Register component types and the React component to render them.
    /// Called from JS via the native module before launching conversation.
    func register(componentTypes: [String], reactComponentName: String) {
        self.registeredTypes = Set(componentTypes)
        self.reactComponentName = reactComponentName
    }

    /// Clear all registrations
    func reset() {
        registeredTypes.removeAll()
        reactComponentName = ""
    }

    var isRegistered: Bool {
        !registeredTypes.isEmpty && !reactComponentName.isEmpty
    }

    // MARK: - AgentforceViewProviding

    func canHandle(type: String) -> Bool {
        registeredTypes.contains(type)
    }

    @MainActor
    func view(for type: String, data: [String: Any]) -> AnyView {
        let props: [String: Any] = [
            "definition": type,
            "properties": data,
        ]
        return AnyView(ReactNativeViewWrapper(
            bridge: bridge,
            moduleName: reactComponentName,
            initialProperties: props
        ))
    }
}

// MARK: - SwiftUI wrapper for RCTRootView

/// Wraps an RCTRootView in a UIViewRepresentable for use in SwiftUI.
private struct ReactNativeViewWrapper: UIViewRepresentable {
    let bridge: RCTBridge?
    let moduleName: String
    let initialProperties: [String: Any]

    func makeUIView(context: Context) -> UIView {
        guard let bridge = bridge else {
            let label = UILabel()
            label.text = "Bridge unavailable"
            label.textAlignment = .center
            return label
        }
        let rootView = RCTRootView(
            bridge: bridge,
            moduleName: moduleName,
            initialProperties: initialProperties
        )
        rootView.backgroundColor = .clear
        return rootView
    }

    func updateUIView(_ uiView: UIView, context: Context) {
        // RCTRootView handles its own updates via the bridge
    }
}

File: AgentforceSDK-ReactNative-Bridge/android/src/main/java/com/salesforce/android/reactagentforce/providers/BridgeViewProvider.kt
/*
 * Copyright (c) 2024-present, salesforce.com, inc.
 * All rights reserved.
 *
 * Bridges the native AgentforceViewProvider interface to React Native.
 * When enabled, delegates rendering of specified component types to a
 * registered React Native component via ReactRootView.
 */
package com.salesforce.android.reactagentforce.providers

import android.os.Bundle
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import com.facebook.react.ReactApplication
import com.facebook.react.ReactRootView
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.salesforce.android.agentforcesdk.components.models.AgentforceComponent
import com.salesforce.android.agentforcesdk.components.models.AgentforceViewProvider
import java.util.concurrent.ConcurrentHashMap

/**
 * Implements AgentforceViewProvider by delegating to a React Native component.
 * Component types are registered synchronously from JS; rendering uses ReactRootView.
 */
class BridgeViewProvider(
    private val reactContext: ReactApplicationContext
) : AgentforceViewProvider {

    /**
     * Maps component definition strings to their React Native component names.
     * Thread-safe: `canHandle` may be called from the SDK rendering thread while
     * `register`/`reset` are called from the JS thread.
     */
    private val componentMap: ConcurrentHashMap<String, String> = ConcurrentHashMap()

    /** Register a 1:1 mapping of component types to React component names. */
    fun register(componentMap: Map<String, String>) {
        this.componentMap.clear()
        this.componentMap.putAll(componentMap)
    }

    /** Clear all registrations */
    fun reset() {
        componentMap.clear()
    }

    val isRegistered: Boolean
        get() = componentMap.isNotEmpty()

    // region AgentforceViewProvider

    override fun canHandle(definition: String): Boolean {
        return componentMap.containsKey(definition)
    }

    @Composable
    override fun GetView(modifier: Modifier, view: AgentforceComponent) {
        val moduleName = componentMap[view.definition] ?: return
        val props = componentToBundle(view)
        AndroidView(
            modifier = modifier,
            factory = { context ->
                ReactRootView(context).apply {
                    val reactApp = reactContext.applicationContext as? ReactApplication
                    val instanceManager = reactApp?.reactNativeHost?.reactInstanceManager
                    startReactApplication(instanceManager, moduleName, props)
                }
            },
            onRelease = { view -> view.unmountReactApplication() }
        )
    }

    /** Convert AgentforceComponent to a Bundle for React Native initial properties */
    private fun componentToBundle(component: AgentforceComponent): Bundle {
        return Bundle().apply {
            putString("definition", component.definition)
            component.name?.let { putString("name", it) }
            putBundle("properties", mapToBundle(component.properties))
            if (component.subComponents.isNotEmpty()) {
                val subArray = component.subComponents.map { componentToBundle(it) }.toTypedArray()
                putParcelableArray("subComponents", subArray)
            }
        }
    }

    /** Recursively convert a Map to a Bundle */
    private fun mapToBundle(map: Map<String, Any>): Bundle {
        return Bundle().apply {
            for ((key, value) in map) {
                when (value) {
                    is String -> putString(key, value)
                    is Int -> putInt(key, value)
                    is Long -> putLong(key, value)
                    is Double -> putDouble(key, value)
                    is Float -> putFloat(key, value)
                    is Boolean -> putBoolean(key, value)
                    is Map<*, *> -> {
                        @Suppress("UNCHECKED_CAST")
                        putBundle(key, mapToBundle(value as Map<String, Any>))
                    }
                    is List<*> -> {
                        // Coerces all elements to strings via toString(). Numeric/nested
                        // type information is lost — acceptable for the current SDK contract
                        // where list properties are display-only string sequences.
                        putStringArray(key, value.map { it?.toString() ?: "" }.toTypedArray())
                    }
                    else -> putString(key, value.toString())
                }
            }
        }
    }

    // endregion
}
